lib/api/apiUtils/object/coldStorage.js

```diff
@@ -28,6 +33,99 @@ function getAmzRestoreResHeader(objMD) {
     return undefined;
 }
 
+
+/**
+ * Check if restore can be done.
+ *
+ * @param {BucketInfo} bucketMD - bucket instance
+ * @param {ObjectMD} objectMD - object metadata
+ * @return {ArsenalError|boolean} - undefined if the conditions for RestoreObject are fulfilled
+ */
+function _validateStartRestore(bucketMD, objectMD) {
+    const isLocationCold = locationConstraints[bucketMD.getLocationConstraint()]?.isCold;
+    if (!isLocationCold) {
+        // return InvalidObjectState error if the bucket of the object is not in cold storage,
+        // not in cold storage means either location cold flag not exists or cold flag is explicit false
+        return errors.InvalidObjectState;
+    }
+    if (objectMD.archive?.restoreCompletedAt
+        && new Date(objectMD.archive?.restoreWillExpireAt) < new Date(Date.now())) {
+        // return InvalidObjectState error if the restored object is expired
+        // but restore info md of this object has not yet been cleared
+        return errors.InvalidObjectState;
+    }
+    if (objectMD.archive?.restoreRequestedAt && !objectMD.archive?.restoreCompletedAt) {
+        // return RestoreAlreadyInProgress error if the object is currently being restored
+        // check if archive.restoreRequestAt exists and archive.restoreCompletedAt not yet exists
+        return errors.RestoreAlreadyInProgress;
+    }
+    return undefined;
+}
+
+/**
+ * Check if the object is already restored
+ *
+ * @param {ObjectMD} objectMD - object metadata
+ * @return {boolean} - true if the object is already restored
+ */
+function isObjectAlreadyRestored(objectMD) {
+    //  check if restoreCompletedAt field exists
+    //  and archive.restoreWillExpireAt > current time
+    return objectMD.archive?.restoreCompletedAt
+        && new Date(objectMD.archive?.restoreWillExpireAt) >= new Date(Date.now());
+}
+
+/**
+ * update restore expiration date.
+ *
+ * @param {BucketInfo} bucketMD - bucket instance
+ * @param {ObjectMD} objectMD - objectMD instance
+ * @param {object} restoreParam - restore param
+ * @param {object} log - werelogs logger
+ * @return {ArsenalError|undefined} internal error if object MD is not valid
+ *
+ */
+function _updateRestoreInfo(bucketMD, objectMD, restoreParam, log) {
+    // eslint-disable-next-line no-param-reassign
+    objectMD.archive.restoreRequestedAt = new Date();
```

Comment:
I assume objectMD.archive will always be defined since `objectMD.archive.archiveInfo` will be set during the archival process. However, I would make sure `Cannot set properties of undefined` error would not be thrown if `objectMD.archive` happens to be undefined.